pointpats/centrography.py

```diff
@@ -825,6 +830,159 @@ def _(points: np.ndarray) -> NDArray[np.float64]:
+    Returns
+    -------
+    a numpy array aligned with the input points that is "True" when a point remains
+    in the dataset, and "False" when the point has been pared.
+
+    Notes
+    -----
+    If convex=False, the concave hull algorithm provided by geopandas is used by default.
+    If you use convex=False, you should also provide a "ratio" argument, which controls
+    the percentage of the input dataset that is contained within the concave hull in
+    each iteration. Generally, ratio should be larger than p if peeling=True.
+    """
+    if isinstance(points, GeoSeries):
+        pass
+    elif isinstance(points, GeoDataFrame):
+        points = points.geometry
+    else:
+        if points.shape[1] == 2:
+            points = GeoSeries(points_from_xy(*points.T))
+        else:
+            points = GeoSeries(points)
+    n_points = len(points)
+    trimmed = pandas.Series(np.zeros(n_points), index=points.index).astype(bool)
+    if hull=='convex':
+        huller = lambda x: x.convex_hull
+    elif hull == 'concave':
+        huller = lambda x: GeoSeries(x).concave_hull(**hull_args).item()
+    elif callable(hull):
+        huller = lambda x: hull(x, **hull_args)
+    else:
+        raise ValueError("hull argument must be either `concave', `convex', or callable.")
+    while True:
+        pointset = points[~trimmed].union_all()
+        hull_ = huller(pointset)
+        on_hull = hull_.boundary.intersects(points)
+        if not on_hull.any():
+            raise ValueError("Hull algorithm must return shape that intersects at least some of the input points.")
+        if peeling:
+            q = (trimmed | on_hull).mean()
+        else:
+            q = ((trimmed).sum() + 1)/n_points
+        if (1 - q) < p:
+            break
+        if peeling:
+            trimmed[on_hull] = True
+        else:
+            # roll the angles by 1 because the angles are aligned on the LEFT EDGE. So, the first
+            # element of get_angles() is the angle centered on 2, with 1 as its left edge.
+            angles = np.roll(shape.get_angles(hull_.boundary), 1)
+            if (angles < 0).any() | (angles > np.pi).any():
+                raise NotImplementedError("negative angle or reflexive angle encountered. Your polygon likely has an incorrect winding direction.")
+            # ring is always closed at the same point as it starts
+            ordered_hull_points = GeoSeries(points_from_xy(*hull_.boundary.xy)[:-1])
+            hull_locs, points_hull_locs = points[on_hull.values].sindex.query(ordered_hull_points, predicate='intersects')
+            points_hull_ixs = points[on_hull.values].index[points_hull_locs]
+            current_knob_on_hull = angles.argmin()
+            current_knob_ix = points_hull_ixs[current_knob_on_hull]
+            # for the points on the hull, get the ith observation's full dataset index
+            trimmed.loc[current_knob_ix] = True
+    return ~trimmed
```

Comment:
Another possible enhancement you might consider is optionally returning the sequence of hulls generated during the trimming process. A long while back, I explored the geometric and spatial properties of these nested hulls, and found that they can reveal interesting patterns in the point set structure.  Never returned to this, but this excellent implementation reminded me of that. Capturing the intermediate hulls would make it possible to later analyze, visualize, or summarize the shape evolution of the dataset as it's pared down. This could be done with a simple flag like`return_hulls=False`, and if enabled, the function could return a tuple `(mask, hulls)`.

Reply:
+1 on this.